### ISSUE TITLE
[Feature request] Partial linking

### ISSUE THREAD

**Issue by aaaaaa123456789** [rednex/rgbds#223]
Part of a reocurring issue we've encountered in different situations is the fact that labels aren't constants, and thus cannot be treated as such in expressions. The difference between two labels is only well-defined for non-fixed section labels only if they belong to the same section, too.

A good solution to every "label is not constant" issue I've found so far (including some longstanding problems, such as `ld a, [hFoo]` taking 3 bytes if `hFoo` is an HRAM label and not a true constant) would be to be able to link different parts of the ROM in sequence, by section type, and then feed the results (and thus the exported symbols) to subsequent assembler and linker passes.

In other words, what I'm proposing is to be able to tell rgbasm/rgblink to only assemble and link sections of a certain type (ROM0, ROMX, WRAM0, WRAMX, and so on), and to be able to use the resulting exported symbols as constants in further rgbasm/rgblink passes.

**Comment by AntonioND** [comment]
I can't think of any way of doing this that doesn't involve a rewrite of a huge part of the toolchain and changing how it fundamentally works.

PS: I guess that for the HRAM loads you are using `ldh`, right?

**Comment by aaaaaa123456789** [comment]
We use constants, the same way pokecrystal does (check hram.asm in that
repo). It mostly works, but it's suboptimal.

While this feature would be incredibly useful (for macros too), I
understand it is not a simple one to implement. Consider it a long-term
goal or something?

**Comment by AntonioND** [comment]
In any case, the problem with "`ld a, [hFoo]` taking 3 bytes if hFoo is an HRAM label and not a true constant" is a different one: rgblink can't change the size of an instruction, it can only patch values. It's the same problem as https://github.com/rednex/rgbds/issues/133

**Comment by aaaaaa123456789** [comment]
While that's true, partial linking would solve the issue anyway. If HRAM is compiled and linked, and the result is fed back to rgbasm when compiling ROM0/ROMX, rgbasm can already know the value of `hFoo` and treat it as a constant, emitting the two-byte version. Hence the final comment in my original message:

>and to be able to use the resulting exported symbols as constants in further rgbasm/rgblink passes.

**Comment by AntonioND** [comment]
I'm going to reject this request. Implementing this would mean that the assembler needs to be able to read object files, not only write them. That is, at least, _weird_. Not only that, it needs to use them as additional input of data that the source code can use, which would mean basically copying code from rgblink into rgbasm, and checking which labels can be safely used because they are constant or not... At that point, it doesn't even make sense to have different binaries, it would make more sense to just have a compiler that can do everything, like gcc.

In short, the value that this adds is really small, it involves a huge rework of rgbasm, and there are not-too-bad ways to do it without it.
